src/App.tsx

```diff
@@ -1,24 +1,219 @@
+      setIsError(true);
+      setIsTodoLoaded(false);
+      setErrorMessage('Data couldn\'t be loaded from the server');
+    }
+  };
+
+  useEffect(() => {
+    getTodosFromServer();
+  }, []);
+
+  const createTodoOnServer = async (query: string) => {
+    const data = {
+      title: query,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({
+      ...data,
+      id: 0,
+    });
+
+    try {
+      setIsTodoLoaded(true);
+      const newTodo = await createTodo(USER_ID, data);
+
+      setTodos([...todos, newTodo]);
+    } catch (error) {
+      setIsError(true);
+      setErrorMessage('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+      setIsTodoLoaded(false);
+    }
+  };
+
+  const removeTodoOnServer = async (id: number) => {
+    try {
+      setIsTodoLoaded(true);
+      await deleteTodo(id);
+      setTodos(todos.filter(todo => todo.id !== id));
+    } catch (error) {
+      setIsError(true);
+      setErrorMessage('Unable to delete a todo');
+    } finally {
+      setTempTodo(null);
+      setIsTodoLoaded(false);
+    }
+  };
+
+  const updateTodoOnServer = async (todo: Todo, isChange?: boolean) => {
+    const todoId = todo.id;
+    const todoCopy = {
+      ...todo,
+    };
+
+    if (isChange) {
+      todoCopy.completed = !todo.completed;
+    }
+
+    try {
+      setIsTodoLoaded(true);
+      await updateTodo(todoId, todoCopy);
+      await getTodosFromServer();
+    } catch (error) {
+      setIsError(true);
+      setErrorMessage('Unable to update a todo');
+    } finally {
+      setTempTodo(null);
+      setIsTodoLoaded(false);
+    }
+  };
+
+  const updateAllTodos = async (todosFromServer: Todo[]) => {
+    const areAllCompleted = todosFromServer.every((todo) => todo.completed);
+
+    try {
+      const updatedTodos = await
+      Promise.all(todosFromServer.map(async (todo) => {
+        if (todo.completed === areAllCompleted) {
+          await patchTodoStatus(todo.id, { completed: !areAllCompleted });
+        }
+
+        return { ...todo, completed: !areAllCompleted };
+      }));
+
+      setTodos(updatedTodos);
+    } catch (error) {
+      setErrorMessage('Unable to update all todos');
+    }
+  };
+
+  const clearCompletedTodo = async () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    try {
+      setIsTodoLoaded(true);
+      await
+      Promise.all(completedTodos.map(todo => deleteTodo(todo.id)));
+      setTodos(todos.filter(todo => !todo.completed));
+    } catch (error) {
+      setIsError(true);
+      setErrorMessage('Unable to clear completed todos');
+    } finally {
+      setIsTodoLoaded(false);
+    }
+  };
+
+  const visibleTodos: Todo[] = useMemo(() => {
+    switch (filterBy) {
+      case FilteredBy.ACTIVE:
+        return todos.filter(todo => !todo.completed);
+
+      case FilteredBy.COMPLETED:
+        return todos.filter(todo => todo.completed);
+
+      case FilteredBy.ALL:
+        return [...todos];
```

Comment:
No need to create copy
```suggestion
        return todos;
```